### PR TITLE
check for datetime+period addition

### LINKED ISSUE #17983: Series[Period] and Timestamp binop causes segfault
No time to look atm:

```
In [1]: import pandas as pd

In [2]: pd.Timestamp('2017') + pd.Series([pd.Period('2017', freq='M')])
Fatal Python error: Cannot recover from stack overflow.

Thread 0x00007000053d1000 (most recent call first):
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 295 in wait
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 551 in wait
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/core/history.py", line 829 in run
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-24>", line 2 in run
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 884 in _bootstrap

Current thread 0x00007fff943f83c0 (most recent call first):
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py", line 78 in <lambda>
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py", line 713 in <lambda>
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py", line 713 in safe_na_op
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py", line 742 in wrapper
  File "<ipython-input-2-5cbafe935254>", line 1 in <module>
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2862 in run_code
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2808 in run_ast_nodes
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2698 in run_cell
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 471 in interact
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/terminal/interactiveshell.py", line 480 in mainloop
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/terminal/ipapp.py", line 356 in start
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/traitlets/config/application.py", line 658 in launch_instance
  File "/Users/taugspurger/Envs/pandas-dev/lib/python3.6/site-packages/IPython/__init__.py", line 125 in start_ipython
  File "/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython", line 11 in <module>
```

The other way hits a maximum recursion error:

```python
In [1]: import pandas as pd

In [2]: pd.Series([pd.Period('2017', freq='M')]) + pd.Timestamp('2017')


---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py in safe_na_op(lvalues, rvalues)
    698             with np.errstate(all='ignore'):
--> 699                 return na_op(lvalues, rvalues)
    700         except Exception:

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py in na_op(x, y)
    674         try:
--> 675             result = expressions.evaluate(op, str_rep, x, y, **eval_kwargs)
    676         except TypeError:

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/computation/expressions.py in evaluate(op, op_str, a, b, use_numexpr, **eval_kwargs)
    203     if use_numexpr:
--> 204         return _evaluate(op, op_str, a, b, **eval_kwargs)
    205     return _evaluate_standard(op, op_str, a, b)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/computation/expressions.py in _evaluate_numexpr(op, op_str, a, b, truediv, reversed, **eval_kwargs)
    118     if result is None:
--> 119         result = _evaluate_standard(op, op_str, a, b)
    120

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/computation/expressions.py in _evaluate_standard(op, op_str, a, b, **eval_kwargs)
     63     with np.errstate(all='ignore'):
---> 64         return op(a, b)
     65

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.Timestamp.__radd__()

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/period.pyx in pandas._libs.period._Period.__add__()

... last 2 frames repeated, from the frame below ...

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.Timestamp.__radd__()

RecursionError: maximum recursion depth exceeded while calling a Python object

During handling of the above exception, another exception occurred:

RecursionError                            Traceback (most recent call last)
<ipython-input-2-18743c02baa6> in <module>()
----> 1 pd.Series([pd.Period('2017', freq='M')]) + pd.Timestamp('2017')

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py in wrapper(left, right, name, na_op)
    736                 lvalues = lvalues.values
    737
--> 738         result = wrap_results(safe_na_op(lvalues, rvalues))
    739         return construct_result(
    740             left,

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py in safe_na_op(lvalues, rvalues)
    707                 if is_object_dtype(lvalues):
    708                     return libalgos.arrmap_object(lvalues,
--> 709                                                   lambda x: op(x, rvalues))
    710             raise
    711

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/algos_common_helper.pxi in pandas._libs.algos.arrmap_object()

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/ops.py in <lambda>(x)
    707                 if is_object_dtype(lvalues):
    708                     return libalgos.arrmap_object(lvalues,
--> 709                                                   lambda x: op(x, rvalues))
    710             raise
    711

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.Timestamp.__radd__()

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/period.pyx in pandas._libs.period._Period.__add__()

... last 2 frames repeated, from the frame below ...

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.Timestamp.__radd__()

RecursionError: maximum recursion depth exceeded while calling a Python object

```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #17983
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/scalar/test_period.py [hunk 1]
```diff
@@ -1038,6 +1038,24 @@ def test_add_raises(self):
...
+    boxes = [lambda x: x, lambda x: pd.Series([x]), lambda x: pd.Index([x])]
+
+    @pytest.mark.parametrize('lbox', boxes)
+    @pytest.mark.parametrize('rbox', boxes)
+    def test_add_timestamp_raises(self, rbox, lbox):
+        # GH # 17983
+        ts = pd.Timestamp('2017')
+        per = pd.Period('2017', freq='M')
+
+        with pytest.raises(TypeError):
```
Comment (gfyoung) [inline pandas/tests/scalar/test_period.py]: Let's make sure that the right error message is hit (use `tm.assert_raises_regex` to be safe, unclear ATM whether we're going to leverage the most recent `pytest` API).

**Comment by gfyoung** [comment]
@jbrockmendel : Don't forget the `whatsnew` also :smile:

File: pandas/tests/scalar/test_period.py [hunk 2]
```diff
@@ -1038,6 +1038,24 @@ def test_add_raises(self):
...
+            lbox(ts) + rbox(per)
+
+        with pytest.raises(TypeError):
+            lbox(per) + rbox(ts)
```
Comment (jreback) [inline pandas/tests/scalar/test_period.py]: can you add a similar test with Series of periods & Timestamp; I think these are now in pandas/tests/series/test_timestamp.py IIRC (maybe make a test_period.py  in series ok too).

**Review by jreback — CHANGES_REQUESTED** [review]
needs a whatsnew note (0.22.0)

File: pandas/tests/scalar/test_period.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/test_period.py]: woops, just pushed without addressing this.  Do you mean a Series that mixed, contains both Periods and Timestamps?

File: pandas/tests/scalar/test_period.py [hunk 3]
```diff
@@ -1038,6 +1038,29 @@ def test_add_raises(self):
...
+               r"ufunc add cannot use operands")
+        with tm.assert_raises_regex(TypeError, msg):
+            lbox(ts) + rbox(per)
+
+        with tm.assert_raises_regex(TypeError, msg):
+            lbox(per) + rbox(ts)
+
+        with tm.assert_raises_regex(TypeError, msg):
+            lbox(per) + rbox(per)
+
```
Comment (jreback) [inline pandas/tests/scalar/test_period.py]: can you add similar tests for sub

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/test_period.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/test_period.py]: OK for todo list?  I'd like to keep this narrow, busy day ahead.

Comment (jreback) [reply pandas/tests/scalar/test_period.py]: ok

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
doesn't seem to fix: https://github.com/pandas-dev/pandas/pull/18205 but maybe helps

**Comment by jbrockmendel** [comment]
> doesn't seem to fix: #18205 but maybe helps

Unrelated.  But that's high on my personal list, along with getting DateOffsets to be immutable.

**Pull request merged**